### PR TITLE
Deprecate SessionContext::create_logical_plan (#4617)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4617

# Rationale for this change

This is part of moving query execution off `SessionContext` onto `DataFrame`, ensuring that a query is executed against the same `SessionState` that planned it. It is also in many cases significantly less code.

This also will drastically reduce the churn from #4607, as large portions of that PR were related to `create_physical_plan` becoming async.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:266]: This somewhat relates to #4607 but also serves avoids issues due to the interior mutability of CatalogList

File: benchmarks/src/bin/nyctaxi.rs
Comment (tustvold) [inline benchmarks/src/bin/nyctaxi.rs:129]: As ctx is interior mutable this would effectively plan, optimize, and execute potentially against at least 3 different states :scream:

File: datafusion/core/tests/sql/subqueries.rs
Comment (tustvold) [inline datafusion/core/tests/sql/subqueries.rs:257]: These map_err were only adding a context of `"error"` so I just opted to remove them :sweat_smile:

File: datafusion/core/src/dataframe.rs
Comment (tustvold) [inline datafusion/core/src/dataframe.rs:742]: I plan to move this off `LogicalPlan` and onto `LogicalPlanBuilder` in a subsequent PR, to be consistent with the rest of the system

File: datafusion/core/tests/sql/errors.rs
Comment (tustvold) [inline datafusion/core/tests/sql/errors.rs:138]: The new API prevents this sort of issue

File: datafusion/core/src/dataframe.rs
Comment (alamb) [inline datafusion/core/src/dataframe.rs:513]: 👍 
I wonder if we should call it `into_unoptimized_plan` for consistency 🤔

File: datafusion/core/src/dataframe.rs
Comment (alamb) [inline datafusion/core/src/dataframe.rs:527]: Is there any reason to make this consume `self` ? Perhaps it could be

```suggestion
    pub fn optimized_plan(&self) -> Result<LogicalPlan> {
```

File: datafusion/core/tests/sql/avro.rs
Comment (alamb) [inline datafusion/core/tests/sql/avro.rs:124]: 🙈  -- this pattern is very old in the DataFusion codebase -- I think it was simply copied around and has never been cleaned up. Thank you 🙏

File: datafusion/core/tests/sql/mod.rs
Comment (alamb) [inline datafusion/core/tests/sql/mod.rs:1076]: This change appears to have lost the check  logical schema and optimized schema are the same. I think that is a valuable check to have in tests

File: datafusion/core/tests/sql/select.rs
Comment (alamb) [inline datafusion/core/tests/sql/select.rs:1301]: I agree that the 20 output row check is covered by `assert_batches_eq` 👍

**Review by alamb — APPROVED** [review]
Thank you @tustvold . I think this is a great change ❤️ 

I really like how this is improving the APIs to make them more consistent and slowly moving us towards a more consistent design.

cc @andygrove @mingmwang @xudong963 @yahoNanJing @liukun4515 

I think we should leave this open for at least another 24 hours to give others a chance to review if they want or note they would like more time.

**Review by xudong963 — COMMENTED** [review]
The title is "Deprecate `SessionContext::create_physical_plan`", but If I'm not mistaken, what is actually discarded is `SessionContext::create_logical_plan`?

**Comment by tustvold** [comment]
Yup, got ahead of myself 😅

File: datafusion/core/src/dataframe.rs
Comment (tustvold) [reply datafusion/core/src/dataframe.rs:527]: I eventually hope to make optimisation consuming, so this is just laying the groundwork for that - https://github.com/apache/arrow-datafusion/issues/4627

**Review by andygrove — APPROVED** [review]
I don't have time to review in detail but LGTM.

File: datafusion/core/src/dataframe.rs
Comment (tustvold) [reply datafusion/core/src/dataframe.rs:513]: Done

**Comment by tustvold** [comment]
Unless there any objections I intend to merge this in the next few hours, hopefully before it ends up with merge conflicts again :sweat_smile:

**Pull request merged**
